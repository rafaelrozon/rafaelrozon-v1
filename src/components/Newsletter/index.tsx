import * as React from 'react';
import * as Yup from 'yup';
import { Formik, Form as FormikForm } from 'formik';
import {
    Label,
    Input,
    Username,
    Row,
    Col,
    SubscribeTo,
    SubscribeWarning,
    Button,
    Message,
    EmailRow,
    InputErrorMessage,
    Container,
} from './styles';

const SignupSchema = Yup.object().shape({
    firstName: Yup.string()
        .min(2, 'Too Short!')
        .max(50, 'Too Long!')
        .required('Required'),
    lastName: Yup.string()
        .min(2, 'Too Short!')
        .max(50, 'Too Long!')
        .required('Required'),
    email: Yup.string().email('Invalid email').required('Required'),
});

function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

enum ErrorTypes {
    GENERIC = 'GENERIC',
}

interface NewsletterError {
    type: string;
    msg: string;
}

interface MailchimpResponse {
    result: string;
    msg: string;
}

interface Props {
    onSave: (
        email: string,
        fields: object,
        override: string
    ) => Promise<MailchimpResponse>;
}

interface State {
    result: MailchimpResponse | null;
    error: NewsletterError | null;
}

class Newsletter extends React.Component<Props, State> {
    state: State = {
        result: null,
        error: null,
    };

    signUpUser = async ({ firstName, lastName, email }) => {
        try {
            const { onSave } = this.props;
            const result = await onSave(
                email,
                {
                    FNAME: capitalizeFirstLetter(firstName),
                    LNAME: capitalizeFirstLetter(lastName),
                },
                null
            );

            this.setState({ result });
            return true;
        } catch (error) {
            this.setState({
                error: {
                    type: ErrorTypes.GENERIC,
                    msg: 'Sorry, some error happened.',
                },
            });
            return false;
        }
    };

    render() {
        const { result, error } = this.state;
        const newsletterError = result && result.result === 'error';
        const newsletterSuccess = result && result.result === 'success';
        const msg = result && result.msg;

        return (
            <Container data-testid="newsletter">
                <SubscribeTo>Subscribe to my Newsletter</SubscribeTo>
                <SubscribeWarning>
                    I won't send you spam. Unsubscribe at any time.
                </SubscribeWarning>

                {error && error.msg && (
                    <Message data-testid="generic-error" type="error">
                        {error.msg}
                    </Message>
                )}

                <Formik
                    initialValues={{
                        firstName: '',
                        lastName: '',
                        email: '',
                    }}
                    onSubmit={this.signUpUser}
                    validationSchema={SignupSchema}
                >
                    <FormikForm>
                        {newsletterError && (
                            <Message
                                data-testid="newsletter-error"
                                type="error"
                                dangerouslySetInnerHTML={{ __html: msg }}
                            />
                        )}
                        {newsletterSuccess && (
                            <Message
                                data-testid="newsletter-success"
                                type="success"
                            >
                                {msg}
                            </Message>
                        )}
                        <Row>
                            <Username>
                                <Col col={6}>
                                    <Label labelFor="firstName">
                                        First Name
                                    </Label>
                                    <Input
                                        id="firstName"
                                        data-testid="firstName"
                                        type="text"
                                        name="firstName"
                                    />
                                    <InputErrorMessage
                                        component="div"
                                        name="firstName"
                                    />
                                </Col>
                                <Col col={6}>
                                    <Label labelFor="lastName">Last Name</Label>
                                    <Input
                                        id="lastNamee"
                                        data-testid="lastName"
                                        type="text"
                                        name="lastName"
                                    />
                                    <InputErrorMessage
                                        component="div"
                                        name="lastName"
                                    />
                                </Col>
                            </Username>
                        </Row>
                        <EmailRow>
                            <Label labelFor="email">Email</Label>
                            <Input
                                id="email"
                                data-testid="email"
                                type="email"
                                name="email"
                            />
                            <InputErrorMessage component="div" name="email" />
                        </EmailRow>
                        <Row>
                            <Col>
                                <Button
                                    data-testid="submit"
                                    type="submit"
                                    value="Subscribe"
                                />
                            </Col>
                        </Row>
                    </FormikForm>
                </Formik>
            </Container>
        );
    }
}

export default Newsletter;
