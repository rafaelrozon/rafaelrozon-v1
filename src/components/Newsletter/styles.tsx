import styled from 'styled-components';
import { Field, ErrorMessage } from 'formik';

export const Container = styled.div`
    margin-top: 3em;
`;

export const SubscribeTo = styled.h2`
    text-align: center;
    margin: 1em 0;
    font-size: 18px;
    font-weight: bold;
    font-family: ${props => props.theme.color.header};
`;

export const SubscribeWarning = styled.p`
    font-size: 0.9em;
    text-align: center;
    margin-bottom: 2em;
`;

// Form
export const Label = styled.label`
    display: block;
    margin-bottom: 0.25em;
    font-size: 0.8em;
`;

export const Input = styled(Field)`
    box-sizing: border-box;
    width: 100%;
    border-radius: 5px;
    border: 1px solid ${props => props.theme.color.darkGreen};
    font-size: 1em;
    line-height: 1.5em;
    padding: 0.6em;
    font-family: ${props => props.theme.fontFamily.mono};
`;

export const Message = styled.p`
    font-weight: bold;
    background: ${props =>
        props.type === 'error'
            ? props.theme.color.orange
            : props.theme.color.lightGreen};
    padding: 1em 0.75em;
    border-radius: 5px;
    color: ${props => props.theme.color.white};
`;

export const Username = styled.div`
    display: flex;
    justify-content: space-evenly;
`;

export const Button = styled.input`
    background: white;
    border: 1px solid #126872;
    padding: 0.5em 1em;
    border-radius: 5px;
    font-size: 1em;
    line-height: 1.5em;

    &:hover {
        opacity: 0.8;
    }
`;

export const InputErrorMessage = styled(ErrorMessage)`
    color: ${props => props.theme.color.mediumOrange};
    margin: 0.5em 0;
`;

export const EmailRow = styled.div`
    margin-bottom: 1em;
`;

// Layout
export const Row = styled.div`
    margin-bottom: 0.8em;
`;

export const Col = styled.div`
    width: ${props => (props.col / 12) * 100 + '%'};
    padding-right: 0.5em;

    &:last-of-type {
        padding-right: 0;
    }
`;
