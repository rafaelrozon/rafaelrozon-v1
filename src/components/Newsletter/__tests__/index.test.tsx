import { render, fireEvent, screen, waitFor } from '../../../utils/testUtils';
import * as React from 'react';
import Newsletter from '..';

const fireChange = (element, value) => {
    fireEvent.change(element, { target: { value } });
}

describe('Newsletter', () => {
    it('should sign user to newsletter', async () => {
        const onSaveMock = jest.fn();
        const firstName = 'bob';
        const lastName = 'john';
        const email = 'bob@gmail.com';
        const { container, getByTestId } = render(
            <Newsletter onSave={onSaveMock} />
        );
        const firstNameInput = getByTestId('firstName');
        const lastNameInput = getByTestId('lastName');
        const emailInput = getByTestId('email');
        const btn = getByTestId('submit');

        await waitFor(() => {
            fireChange(firstNameInput, firstName);
            fireChange(lastNameInput, lastName);
            fireChange(emailInput, email);

            btn.click();
        })

        const payload = {
            FNAME: 'Bob',
            LNAME: 'John',
        };
        expect(onSaveMock).toBeCalledWith(email, payload, null);
        expect(container).toMatchSnapshot();
        expect(firstNameInput).toBeTruthy();
        expect(lastNameInput).toBeTruthy();
        expect(emailInput).toBeTruthy();
        expect(btn).toBeTruthy();
    });

    it('should display error message if api fails', async () => {
        const onSaveMock = jest.fn(async () => ({
            result: 'error',
            msg: 'api failed',
        }));
        const firstName = 'bob';
        const lastName = 'john';
        const email = 'bob@gmail.com';
        const { container, getByTestId } = render(
            <Newsletter onSave={onSaveMock} />
        );

        const firstNameInput = getByTestId('firstName');
        const lastNameInput = getByTestId('lastName');
        const emailInput = getByTestId('email');
        const btn = getByTestId('submit');

        fireChange(firstNameInput, firstName);
        fireChange(lastNameInput, lastName);
        fireChange(emailInput, email);

        getByTestId('submit').click();

        await waitFor(() => {
            expect(getByTestId('newsletter-error')).toBeTruthy();
            expect(container).toMatchSnapshot();
            expect(firstNameInput).toBeTruthy();
            expect(lastNameInput).toBeTruthy();
            expect(emailInput).toBeTruthy();
            expect(btn).toBeTruthy();
        });
    });

    it('should display success message if api is ok', async () => {
        const onSaveMock = jest.fn(async () => ({
            result: 'success',
            msg: 'email saved',
        }));
        const firstName = 'bob';
        const lastName = 'john';
        const email = 'bob@gmail.com';
        const { container, getByTestId } = render(
            <Newsletter onSave={onSaveMock} />
        );

        const firstNameInput = getByTestId('firstName');
        const lastNameInput = getByTestId('lastName');
        const emailInput = getByTestId('email');
        const btn = getByTestId('submit');

        fireChange(firstNameInput, firstName);
        fireChange(lastNameInput, lastName);
        fireChange(emailInput, email);

        getByTestId('submit').click();

        await waitFor(() => {
            expect(getByTestId('newsletter-success')).toBeTruthy();
            expect(container).toMatchSnapshot();
            expect(firstNameInput).toBeTruthy();
            expect(lastNameInput).toBeTruthy();
            expect(emailInput).toBeTruthy();
            expect(btn).toBeTruthy();
        });
    });

    it('should display generic error message', async () => {
        const onSaveMock = jest.fn(() => {throw new Error('it should fail')});
        const firstName = 'bob';
        const lastName = 'john';
        const email = 'bob@gmail.com';
        const { container, getByTestId } = render(
            <Newsletter onSave={onSaveMock} />
        );

        const firstNameInput = getByTestId('firstName');
        const lastNameInput = getByTestId('lastName');
        const emailInput = getByTestId('email');
        const btn = getByTestId('submit');

        fireChange(firstNameInput, firstName);
        fireChange(lastNameInput, lastName);
        fireChange(emailInput, email);

        getByTestId('submit').click();

        await waitFor(() => {
            expect(getByTestId('generic-error')).toBeTruthy();
            expect(container).toMatchSnapshot();
            expect(firstNameInput).toBeTruthy();
            expect(lastNameInput).toBeTruthy();
            expect(emailInput).toBeTruthy();
            expect(btn).toBeTruthy();
        });
    });
});
