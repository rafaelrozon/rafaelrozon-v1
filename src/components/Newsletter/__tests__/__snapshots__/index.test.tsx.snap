// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletter should display error message if api fails 1`] = `
<div>
  <div
    data-testid="newsletter"
  >
    <h2
      class="sc-AxjAm hxvytf"
    >
      Subscribe to my Newsletter
    </h2>
    <p
      class="sc-AxirZ liNIsn"
    >
      I won't send you spam. Unsubscribe at any time.
    </p>
    <form
      action="#"
    >
      <p
        class="sc-AxhUy bOBYiu"
        data-testid="newsletter-error"
        type="error"
      >
        api failed
      </p>
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-AxgMl drUjts"
        >
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              First Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="firstName"
              id="firstName"
              name="firstName"
              type="text"
              value="bob"
            />
          </div>
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              Last Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="lastName"
              id="lastNamee"
              name="lastName"
              type="text"
              value="john"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-AxmLO fNcoXh"
      >
        <label
          class="sc-AxiKw iheSGU"
        >
          Email
        </label>
        <input
          class="sc-AxhCb cAA-drw"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value="bob@gmail.com"
        />
      </div>
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-fzoLsD eTQKtC"
        >
          <input
            class="sc-AxheI hWNLvW"
            data-testid="submit"
            type="submit"
            value="Subscribe"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Newsletter should display generic error message 1`] = `
<div>
  <div
    data-testid="newsletter"
  >
    <h2
      class="sc-AxjAm hxvytf"
    >
      Subscribe to my Newsletter
    </h2>
    <p
      class="sc-AxirZ liNIsn"
    >
      I won't send you spam. Unsubscribe at any time.
    </p>
    <p
      class="sc-AxhUy bOBYiu"
      data-testid="generic-error"
      type="error"
    >
      Sorry, some error happened.
    </p>
    <form
      action="#"
    >
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-AxgMl drUjts"
        >
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              First Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="firstName"
              id="firstName"
              name="firstName"
              type="text"
              value="bob"
            />
          </div>
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              Last Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="lastName"
              id="lastNamee"
              name="lastName"
              type="text"
              value="john"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-AxmLO fNcoXh"
      >
        <label
          class="sc-AxiKw iheSGU"
        >
          Email
        </label>
        <input
          class="sc-AxhCb cAA-drw"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value="bob@gmail.com"
        />
      </div>
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-fzoLsD eTQKtC"
        >
          <input
            class="sc-AxheI hWNLvW"
            data-testid="submit"
            type="submit"
            value="Subscribe"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Newsletter should display success message if api is ok 1`] = `
<div>
  <div
    data-testid="newsletter"
  >
    <h2
      class="sc-AxjAm hxvytf"
    >
      Subscribe to my Newsletter
    </h2>
    <p
      class="sc-AxirZ liNIsn"
    >
      I won't send you spam. Unsubscribe at any time.
    </p>
    <form
      action="#"
    >
      <p
        class="sc-AxhUy fVwsKs"
        data-testid="newsletter-success"
        type="success"
      >
        email saved
      </p>
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-AxgMl drUjts"
        >
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              First Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="firstName"
              id="firstName"
              name="firstName"
              type="text"
              value="bob"
            />
          </div>
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              Last Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="lastName"
              id="lastNamee"
              name="lastName"
              type="text"
              value="john"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-AxmLO fNcoXh"
      >
        <label
          class="sc-AxiKw iheSGU"
        >
          Email
        </label>
        <input
          class="sc-AxhCb cAA-drw"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value="bob@gmail.com"
        />
      </div>
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-fzoLsD eTQKtC"
        >
          <input
            class="sc-AxheI hWNLvW"
            data-testid="submit"
            type="submit"
            value="Subscribe"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Newsletter should sign user to newsletter 1`] = `
<div>
  <div
    data-testid="newsletter"
  >
    <h2
      class="sc-AxjAm hxvytf"
    >
      Subscribe to my Newsletter
    </h2>
    <p
      class="sc-AxirZ liNIsn"
    >
      I won't send you spam. Unsubscribe at any time.
    </p>
    <form
      action="#"
    >
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-AxgMl drUjts"
        >
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              First Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="firstName"
              id="firstName"
              name="firstName"
              type="text"
              value="bob"
            />
          </div>
          <div
            class="sc-fzoLsD bpwvRW"
          >
            <label
              class="sc-AxiKw iheSGU"
            >
              Last Name
            </label>
            <input
              class="sc-AxhCb cAA-drw"
              data-testid="lastName"
              id="lastNamee"
              name="lastName"
              type="text"
              value="john"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-AxmLO fNcoXh"
      >
        <label
          class="sc-AxiKw iheSGU"
        >
          Email
        </label>
        <input
          class="sc-AxhCb cAA-drw"
          data-testid="email"
          id="email"
          name="email"
          type="email"
          value="bob@gmail.com"
        />
      </div>
      <div
        class="sc-fzozJi iKvhFv"
      >
        <div
          class="sc-fzoLsD eTQKtC"
        >
          <input
            class="sc-AxheI hWNLvW"
            data-testid="submit"
            type="submit"
            value="Subscribe"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;
